pulp_2to3_migration/app/pre_migration.py

```diff
@@ -557,27 +579,55 @@ def mark_removed_resources(plan, type_to_repo_ids):
+        removed_dist_object_ids = premigrated_dist_object_ids - mongo_dist_object_ids
+
+        removed_dists = []
+        for pulp2dist in Pulp2Distributor.objects.filter(
+                pulp2_object_id__in=removed_dist_object_ids):
+            pulp2dist.not_in_pulp2 = True
+            removed_dists.append(pulp2dist)
+
+        Pulp2Distributor.objects.bulk_update(objs=removed_dists,
+                                             fields=['not_in_plan'],
+                                             batch_size=1000)
+
 
 def delete_old_resources(plan):
     """
-    Delete pre-migrated and migrated data for resources which are migrated fully on every run.
+    Delete old Publications/Distributions which are no longer present in Pulp2.
+
+    It's critical to remove Distributions to avoid base_path overlap.
+    It make the migration logic easier if we remove old Publications as well.
 
-    Delete pre-migrated distributors, Publications,
-    and Distributions according to a specified Migration Plan.
+    Delete criteria:
+        - pulp2distributor is no longer in plan
+        - pulp2repository content changed (repo.is_migrated=False) or it is no longer in plan
 
     Args:
         plan(MigrationPlan): A Migration Plan
 
     """
-
-    for plugin in plan.get_plugin_plans():
-        with transaction.atomic():
-            for distributor_type, distributor_migrator in \
-                    plugin.migrator.distributor_migrators.items():
-                # delete all pre-migrated distributors for this plugin
-                Pulp2Distributor.objects.filter(pulp2_type_id=distributor_type).delete()
-                # delete all migrated Publications and Distributions for this plugin
-                for pub_model in distributor_migrator.pulp3_publication_models:
-                    pub_model.objects.all().delete()
-                for distribution_model in distributor_migrator.pulp3_distribution_models:
-                    distribution_model.objects.all().delete()
+    repos_with_old_distributions_qs = Pulp2Repository.objects.filter(
+        Q(is_migrated=False) | Q(not_in_plan=True))
+
+    old_dist_query = Q(pulp3_distribution__isnull=False) | Q(pulp3_publication__isnull=False)
+    old_dist_query &= Q(pulp2_repository__in=repos_with_old_distributions_qs) | Q(not_in_plan=True)
+
+    with transaction.atomic():
+        pulp2distributors_with_old_distributions_qs = Pulp2Distributor.objects.filter(
+            old_dist_query).only('pulp3_distribution')
+
+        for pulp2distributor in pulp2distributors_with_old_distributions_qs:
+            if pulp2distributor.is_migrated:
+                pulp2distributor.is_migrated = False
+                pulp2distributor.save()
+            if pulp2distributor.pulp3_publication:
+                pulp2distributor.pulp3_publication.delete()
+            if pulp2distributor.pulp3_distribution:
+                pulp2distributor.pulp3_distribution.delete()
```

Comment:
It's a pain that they are removed one by one.
I'm open for suggestion what to do here.

Reply:
```select_related``` might work.

Reply:
I think I misunderstood what ```select_related``` does.  Anyway, something like this might work:

```
pulp2distributors_with_old_distributions_qs.bulk_update(is_migrated=True)
pulp2distributors_with_old_distributions_qs.pulp3_publication_set.delete()
pulp2distributors_with_old_distributions_qs.pulp3_distribution_set.delete()
```